# Tensor Normalization and Standardization Assingment

This is the PyTorch tensor normalization assingment for the leaky.ai Introduction to Deep Learning with PyTorch course (www.leaky.ai).  

To execute this assingment head over to Google Colab and load the notebook.

DIRECTIONS
1.  Load notebook into colab.research.google.com
2.  Replace the [TBD]'s with your own code
3.  Execute the notebook after completing each cell and check your answers

In [4]:
# Import PyTorch and check the version
# Create a PyTorch tensor with the following content
# Tensor a should contain [[1,2,3,4,5],[6,7,8,9,10],[11,12,13,14,15],[16,17,18,19,20]]
import torch
torch.manual_seed(6)
torch.__version__
a = torch.tensor([[1,2,3,4,5],[6,7,8,9,10],[11,12,13,14,15],[16,17,18,19,20]], dtype=torch.float)
print (a)

tensor([[ 1.,  2.,  3.,  4.,  5.],
        [ 6.,  7.,  8.,  9., 10.],
        [11., 12., 13., 14., 15.],
        [16., 17., 18., 19., 20.]])


# Finding Maximum Values
Here you will explore different approachs to extract the highest values from tensors.

In [2]:
# Find the highest and lowest value in the entire tensor 
maximum = a.max()
minimum = a.min()

print (f"Max: {maximum.item()}\nMin: {minimum.item()}")

Max: 20.0
Min: 1.0


### Correct Answer:  
<pre>
Values:
tensor([[ 5.,  4.],
        [10.,  9.],
        [15., 14.],
        [20., 19.]])

Indicies:
tensor([[4, 3],
        [4, 3],
        [4, 3],
        [4, 3]])
</pre>

# Standardize Columns
For tabular datasets, you will likely need to standardize or normalize the data before using it for training.

In [3]:
# Standardize the column values

# Start by calculating the mean for each column (dim=0)
mean = a.mean(dim=0)

# Then, calculate the standard deviation for each column (dim=0)
std = a.std(dim=0)

# Standardize the data by subtracting the mean and dividing by st. dev.
a_std = (a - mean) / std

# Output the results
print (a)
print (f"\nMeans:\n{mean}")
print (f"\nStandard Deviations:\n{std}")
print (f"\nStandardized:\n{a_std}")

tensor([[ 1.,  2.,  3.,  4.,  5.],
        [ 6.,  7.,  8.,  9., 10.],
        [11., 12., 13., 14., 15.],
        [16., 17., 18., 19., 20.]])

Means:
tensor([ 8.5000,  9.5000, 10.5000, 11.5000, 12.5000])

Standard Deviations:
tensor([6.4550, 6.4550, 6.4550, 6.4550, 6.4550])

Standardized:
tensor([[-1.1619, -1.1619, -1.1619, -1.1619, -1.1619],
        [-0.3873, -0.3873, -0.3873, -0.3873, -0.3873],
        [ 0.3873,  0.3873,  0.3873,  0.3873,  0.3873],
        [ 1.1619,  1.1619,  1.1619,  1.1619,  1.1619]])


### Correct Answer:  
<pre>
tensor([[ 1.,  2.,  3.,  4.,  5.],
        [ 6.,  7.,  8.,  9., 10.],
        [11., 12., 13., 14., 15.],
        [16., 17., 18., 19., 20.]])

Means:
tensor([[ 8.5000,  9.5000, 10.5000, 11.5000, 12.5000]])

Standard Deviations:
tensor([[6.4550, 6.4550, 6.4550, 6.4550, 6.4550]])

Standardized:
tensor([[-1.1619, -1.1619, -1.1619, -1.1619, -1.1619],
        [-0.3873, -0.3873, -0.3873, -0.3873, -0.3873],
        [ 0.3873,  0.3873,  0.3873,  0.3873,  0.3873],
        [ 1.1619,  1.1619,  1.1619,  1.1619,  1.1619]])</pre>